src/main/java/org/apache/maven/shared/archiver/PomPropertiesUtil.java

```diff
@@ -61,34 +59,29 @@ private boolean sameContents(Properties props, Path file) throws IOException {
         return fileProps.equals(props);
     }
 
-    private void createPropertiesFile(Properties properties, Path outputFile, boolean forceCreation)
+    private void createPropertiesFile(Properties unsortedProperties, Path outputFile, boolean forceCreation)
             throws IOException {
         Path outputDir = outputFile.getParent();
         if (outputDir != null && !Files.isDirectory(outputDir)) {
             Files.createDirectories(outputDir);
         }
-        if (!forceCreation && sameContents(properties, outputFile)) {
+        if (!forceCreation && sameContents(unsortedProperties, outputFile)) {
             return;
         }
 
-        try (PrintWriter pw = new PrintWriter(outputFile.toFile(), StandardCharsets.ISO_8859_1.name());
-                StringWriter sw = new StringWriter()) {
-
-            properties.store(sw, null);
-
-            List<String> lines = new ArrayList<>();
-            try (BufferedReader r = new BufferedReader(new StringReader(sw.toString()))) {
-                String line;
-                while ((line = r.readLine()) != null) {
-                    if (!line.startsWith("#")) {
-                        lines.add(line);
-                    }
-                }
-            }
-
-            Collections.sort(lines);
-            for (String l : lines) {
-                pw.println(l);
+        // For reproducible builds, sort the properties and drop comments.
+        // The java.util.Properties class doesn't guarantee order so we have
+        // to write the file using a Writer.
+        Set<String> propertyNames = unsortedProperties.stringPropertyNames();
+        List<String> sortedPropertyNames = new ArrayList<>(propertyNames);
+        Collections.sort(sortedPropertyNames);
+
+        try (Writer out = Files.newBufferedWriter(outputFile, StandardCharsets.ISO_8859_1)) {
+            for (String key : sortedPropertyNames) {
+                out.write(key);
+                out.write(": ");
+                out.write(unsortedProperties.getProperty(key));
```

Comment:
I think that's wrong.  We need escaping here. See https://github.com/openjdk/jdk/blob/8c2b4f62714f26ab3bc4808c734502af632a1eef/src/java.base/share/classes/java/util/Properties.java#L686-L738

Reply:
I'll add this.

Reply:
Now I see why the original code wrote out a properties file and read it back in. That avoided the need to reimplement the escaping.

Reply:
However, that might not always work. It relies on the Properties format being consistent across VMs and Java versions and it might now be. We're accounting for comments, separator character, and ordering here, but there are other possible differences that can occur. It's risky to assume that Eclipse Temurin 21 is going to produce the same output as OpenJDK 8.

Reply:
I don't think it's risky.  That part is very clearly specified. Changing the way properties file are stored would be a big breakage.
See https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#store-java.io.Writer-java.lang.String- for an exact explanation of what is written.
The way key / separator / value are written is clearly specified and has not changed from JDK 8 to JDK 24.

> Then every entry in this Properties table is written out, one per line. For each entry the key string is written, then an ASCII =, then the associated element string. For the key, all space characters are written with a preceding \ character. For the element, leading space characters, but not embedded or trailing space characters, are written with a preceding \ character. The key and element characters #, !, =, and : are written with a preceding backslash to ensure that they are properly loaded.

Feel free to re-implement this mechanism, but I'm not really sure it's worth it.

Reply:
To guarantee the output, I think we are going to need to reimplement the escaping here. This could be tricky. What are the rules about copy-pasting OpenJDK code into an Apache project? We probably can't do that since it's GPL, not ASpache licensed.

Reply:
And also:
> [...] the input/output stream is encoded in ISO 8859-1 character encoding. Characters that cannot be directly represented in this encoding can be written using Unicode escapes as defined in section [3.3 (https://docs.oracle.com/javase/specs/jls/se24/html/jls-3.html#jls-3.3) of The Java Language Specification; only a single 'u' character is allowed in an escape sequence.

Reply:
> To guarantee the output, I think we are going to need to reimplement the escaping here. This could be tricky. What are the rules about copy-pasting OpenJDK code into an Apache project? We probably can't do that since it's GPL, not ASpache licensed.

The output is already guaranteed by the javadoc of the `Properties.store()` method.  I really don't see what additional guarantees you're looking for.